### ISSUE TITLE
async coding shred recovery outside of insert_shred

### ISSUE THREAD

**Issue by sakridge** [solana-labs/solana#28634]
#### Problem

Coding shred recovery is done synchronously inside insert_shred and blocks other shred inserts while it's happening. There may be other data shreds in the pipeline that would obviate the need for even recovering those shreds as well. Coding recovery is quite CPU intensive.

#### Proposed Solution

Asynchronous service which gets fed the conditions for shred recovery from insert_shreds and prioritizes and performs the shred recovery if necessary for a missing set of shreds.

**Comment by steviez** [comment]
@linuskendall - Per our conversation, this is not staffed but you may be interested in playing around with this on your side

**Comment by sakridge** [comment]
cc @yhchiang-sol

**Comment by steviez** [comment]
I had hacked together a branch to get some data on this. The idea is track the source of every data shred (turbine/recovery/repair), and then if we see that data shred come through insertion chain again, update a counter that corresponds to the initial insertion source and the duplicate source.The branch is [here](https://github.com/steviez/solana/tree/dup_shreds_data).

I let this run for about 21 hours against mainnet. The full set can be found [here](https://docs.google.com/spreadsheets/d/1CFexXI5Solvb2dQdxpLftXb-Zs6iySCXWl6ZcB9nrBU/edit#gid=198067710), but I have included the summary inline:

<img width="1036" alt="image" src="https://user-images.githubusercontent.com/5400107/204661847-af0fff20-3cb1-4edf-9cee-a8ffaf860faf.png">

The columns indicate slot, number of unique shreds inserted in each slot, and then the matrix of duplicate insertions across the possible first/subsequent shred source combinations. The percentages use the sum of the number of shreds in the blocks as the denominator.

Thus, it can be said that 38.97% of the data shreds in these blocks showed up from turbine, but were first recovered. This 38.97% number is pretty significant, so it seems there is merit to the theory that we are recovering shreds we don't need to by doing recovery inline with insertion.

Some of the other numbers are interesting as well, but out of scope for this issue.

**Comment by brennanwatt** [comment]
Created a Proof-of-concept that can be found [here](https://github.com/brennanwatt/solana/tree/async_shred_recover).

At a high level, this branch behaves like so:
1. Shred insertion function send erasure set information for sets that saw a shred insertion to a separate recovery thread
2. Recovery thread performs recovery on erasure set (has checks to prevent from duplicate recoveries as well as mechanism for deferring recovery - useful for avoiding mass turbine/recovery shred collisions)
3. Recovery thread sends recovered shreds back to shred insertion function
4. Shred insertion function inserts recovered shreds (checks for duplicates)

Testing these changes on MNB, I am seeing ~100ms reduction in time spent doing "recovery" in the blockstore insertion thread (from ~240ms to ~140ms) per 3 second reporting window. This is because the actual recovery has been punted to a separate thread, so this time essentially represents recovered shred insertion time. The recovery thread is busy ~10% of the time.

**Comment by steviez** [comment]
Good work, this looks quite promising! One other metric we should double check that we aren't negatively impacting is the latency for seeing full slots. `shred_insert_is_full.total_time_ms` captures the duration between first shred and slot full; let's confirm that this number looks comparable.

A situation that comes to mind where the above metric would increase is if we need a shred that we didn't/won't get over turbine, but the shred is recoverable. With async, there will now be some additional delay for when we recover and insert that shred.

**Comment by brennanwatt** [comment]
It does appear that block times are being adversely impacted by delaying the start of recovery. I performed a quick sweep across some delays and am observing the following:
- with vanilla master, observing 385ms block inserted time
- with 10ms delay, observing 440ms block inserted time and ~4.5 shreds recovered per erasure set
- with 5ms delay, observing 410ms block inserted time and ~8 shreds recovered per erasure set
- with 1ms delay, observing 400ms block inserted time and ~10 shreds recovered per erasure set
As expected, the shorter the delay, the more overall recoveries are performed, the more attempted duplicate blockstore insertions we see, and the more CPU time we spend on insertion/recovery

**Comment by brennanwatt** [comment]
I believe the data is telling us that sticking with inline recovery is the better option at this time for the following reasons:
- Inline recovery doesn't actually take that long. Inserting shreds (~70us) seems to take longer than recovering them (~30us assuming we have the metadata handy). We only spend ~3% of our time performing recovery.
- Delaying erasure set recovery delays the time that the full slot will be inserted in the blockstore. Ensuring we feed replay as fast as possible seems more important than saving a few MIPS
- The incoming shred pipeline isn't being neglected for very long due to shred recovery. A shred comes in every ~260us on average, so recovering an entire erasure set (~1ms) would only result in building a backlog of ~4 shreds.

**Comment by steviez** [comment]
> It does appear that block times are being adversely impacted by delaying the start of recovery. I performed a quick sweep across some delays and am observing the following:

Out of curiosity, could you get a measurement for 0ms delay easily? Also, how long did you let it run to collect the 10/5/1 ms delay samples.

Thinking out loud, 0ms wait means we're recovering everything the side thread gets anyways. This would be comparable to doing the recovery inline, except inline removes the overhead of sending things back and forth, and inline removes the latency of having to wait until insert_iteration_x to get the recovered shreds from insert_iteration_x-1.

At the moment, I'm inclined to agree with you that keeping things inline seems to be the move. However, I'd say let's put this on the backburner and see if any ideas hit us in the next couple days. If not, we close it out.

**Comment by brennanwatt** [comment]
> Out of curiosity, could you get a measurement for 0ms delay easily? Also, how long did you let it run to collect the 10/5/1 ms delay samples.
> 
> Thinking out loud, 0ms wait means we're recovering everything the side thread gets anyways. This would be comparable to doing the recovery inline, except inline removes the overhead of sending things back and forth, and inline removes the latency of having to wait until insert_iteration_x to get the recovered shreds from insert_iteration_x-1.
> 
> At the moment, I'm inclined to agree with you that keeping things inline seems to be the move. However, I'd say let's put this on the backburner and see if any ideas hit us in the next couple days. If not, we close it out.

Just collected 'no delay' data. I'm observing 400ms block inserted time.

For collecting the various shred recovery delay samples, I would wait until the validator was caught up and then let it run for ~30 minutes.

You're right that 'no delay' is similar to inline except for the thread comms overhead and when the recovered shreds get inserted. There are some little nuances such as if we have recently inserted shreds data readily available or need to read it from Rocks, but these are minor.

I'm wondering if some of our block delay can be attributed to just waiting for the receiver in `run_insert` so that we can proceed with inserting the recovered shreds. We spend ~2/3 of our time here waiting to receive packets. Should be quick to run an experiment to confirm this impact

**Comment by steviez** [comment]
> UPDATE: I'm seeing a modest improvement when breaking out of the receiver wait when there are pending shreds

Just seeing this as you don't get notifications for comment updates, but is the modest improvement here over the `no delay` case or the `inline` case?

**Comment by brennanwatt** [comment]
The modest improvement is over the 'no delay' case. Still lagging behind the in-line case when it comes to slot full times

**Comment by brennanwatt** [comment]
> Please also note that with current "legacy shreds", we only need to recover data shreds. But with new Merkle shreds we are recovering both code and data shreds. Even if we receive all code and data shreds in the erasure batch from turbine, the chances that all 64 code and data shreds are received before we can locally reconstruct them seems low to me. In that case if we are optimizing for shreds propagation latency then we probably want to run erasure recovery as soon as possible and do not introduce any delay or waiting.

Yep, I agree with this. Optimistically waiting for more turbine shreds hurts propagation latency in practice even if we can save some CPU cycles. The amount of CPU saved is small relative to the impact to propagation latency, so it doesn't seem worth it. Also, once we decide we're going to recover as soon as we can, doing this inline with insertion so we have all the metadata and recently inserted shreds handy also makes sense.

**Comment by steviez** [comment]
> and see if any ideas hit us in the next couple days. If not, we close it out.

I don't have any other ideas of things to try, and echoing the last two comments, inline/immediate recovery is the winner for minimizing shred ingestion latency. It was still an interesting experiment to run (and having the data to drive the decision is better than speculation), but it seems closing the issue with no action is the move forward.
